src/Nethermind/Nethermind.Serialization.Rlp/CompactReceiptStorageDecoder.cs

```diff
@@ -0,0 +1,237 @@
+            {
+                decoderContext.Check(lastCheck);
+            }
+
+            txReceipt.Bloom = new Bloom(txReceipt.Logs);
+
+            return txReceipt;
+        }
+
+        public void DecodeStructRef(scoped ref Rlp.ValueDecoderContext decoderContext, RlpBehaviors rlpBehaviors,
+            out TxReceiptStructRef item)
+        {
+            item = new TxReceiptStructRef();
+
+            if (decoderContext.IsNextItemNull())
+            {
+                decoderContext.ReadByte();
+                return;
+            }
+
+            decoderContext.SkipLength();
+
+            Span<byte> firstItem = decoderContext.DecodeByteArraySpan();
+            if (firstItem.Length == 1)
+            {
+                item.StatusCode = firstItem[0];
+            }
+            else
+            {
+                item.PostTransactionState =
+                    firstItem.Length == 0 ? new KeccakStructRef() : new KeccakStructRef(firstItem);
+            }
+
+            item.Sender = (decoderContext.DecodeAddress() ?? Address.Zero).ToStructRef();
+            item.GasUsedTotal = (long)decoderContext.DecodeUBigInt();
+
+            // Need to regenerate bloom. Can't lazily load log.
+            int lastCheck = decoderContext.ReadSequenceLength() + decoderContext.Position;
+            List<LogEntry> logEntries = new();
+            while (decoderContext.Position < lastCheck)
+            {
+                logEntries.Add(Rlp.Decode<LogEntry>(ref decoderContext, RlpBehaviors.AllowExtraBytes));
+            }
+            item.Logs = logEntries.ToArray();
+
+            item.Bloom = new Bloom(item.Logs).ToStructRef();
+        }
+
+        public Rlp Encode(TxReceipt item, RlpBehaviors rlpBehaviors = RlpBehaviors.None)
+        {
+            RlpStream rlpStream = new(GetLength(item, rlpBehaviors));
+            Encode(rlpStream, item, rlpBehaviors);
+            return new Rlp(rlpStream.Data);
+        }
+
+        public void Encode(RlpStream rlpStream, TxReceipt? item, RlpBehaviors rlpBehaviors = RlpBehaviors.None)
+        {
+            if (item is null)
+            {
+                rlpStream.EncodeNullObject();
+                return;
+            }
+
+            (int totalContentLength, int logsLength) = GetContentLength(item, rlpBehaviors);
+
+            bool isEip658receipts = (rlpBehaviors & RlpBehaviors.Eip658Receipts) == RlpBehaviors.Eip658Receipts;
+
+            // Note: Any byte saved here is about 3GB on mainnet.
+            rlpStream.StartSequence(totalContentLength);
+            if (isEip658receipts)
+            {
+                rlpStream.Encode(item.StatusCode);
+            }
+            else
+            {
+                rlpStream.Encode(item.PostTransactionState);
+            }
+
+            rlpStream.Encode(item.Sender);
+            rlpStream.Encode(item.GasUsedTotal);
+
+            rlpStream.StartSequence(logsLength);
+
+            for (int i = 0; i < item.Logs.Length; i++)
```

Comment:
Hoist `item.Logs` to a local array e.g.

```csharp
LogEntry[] logs = item.Logs;
for (int i = 0; i < logs.Length; i++)
{
    rlpStream.Encode(logs[i]);
}
```

Then the Jit will eliminate all the bounds checking. Otherwise its defensive assuming `item.Logs` can change during the loop so it needs to keep looking up the values and checking `i` hasn't gone out of range.

Reply:
Done